bench/abseil-cpp/optimized/node_hash_map_test.cc.ll

```diff
@@ -128684,19 +128677,19 @@ lpad:                                             ; preds = %call38.i.i.noexc147
   br label %ehcleanup93
 
 if.else:                                          ; preds = %while.body.i.i
-  invoke void @_ZN7testing7MessageC1Ev(ptr noundef nonnull align 8 dereferenceable(8) %ref.tmp4)
```

Comment:
Miscompilation?

Reply:
I think we need to check why the if.else basicblock is unreachable. If it becomes unreachable, this is normal. I will check and let you know.

Reply:
I extracted IRs which involved reaching if.else BB. 
```


%vecinit.i.i.i.i = insertelement <16 x i8> poison, i8 %conv.i.i.i, i64 0
%.pre = load <16 x i8>, ptr getelementptr inbounds ([32 x i8], ptr @_ZN4absl18container_internal11kEmptyGroupE, i64 0, i64 16), align 16, !noalias !2960

%cmp.i.i.i.i94 = icmp eq <16 x i8> %vecinit15.i.i.i.i, %.pre
%8 = bitcast <16 x i1> %cmp.i.i.i.i94 to i16
%cmp.i.not26.i.i = icmp eq i16 %8, 0
 
while.body.i.i:   
 br i1 %cmp.i.not26.i.i, label %for.end.i.i, label %if.else
```

becase "%cmp.i.not26.i.i" become poison we cannot reach %if.else block. so, it seems normal, IMO.

i think this IR will become poison. 
%vecinit.i.i.i.i = insertelement <16 x i8> poison, i8 %conv.i.i.i, i64 0
am I right?

Reply:
I extracted IRs which involved reaching if.else BB. 
```

%vecinit.i.i.i.i = insertelement <16 x i8> poison, i8 %conv.i.i.i, i64 0
%.pre = load <16 x i8>, ptr getelementptr inbounds ([32 x i8], ptr @_ZN4absl18container_internal11kEmptyGroupE, i64 0, i64 16), align 16, !noalias !2960

%vecinit15.i.i.i.i = shufflevector <16 x i8> %vecinit.i.i.i.i, <16 x i8> poison, <16 x i32> zeroinitializer
%cmp.i.i.i.i94 = icmp eq <16 x i8> %vecinit15.i.i.i.i, %.pre
%8 = bitcast <16 x i1> %cmp.i.i.i.i94 to i16
%cmp.i.not26.i.i = icmp eq i16 %8, 0
 
while.body.i.i:   
 br i1 %cmp.i.not26.i.i, label %for.end.i.i, label %if.else
 
 
```

becase "%cmp.i.not26.i.i" become poison we cannot reach %if.else block. so, it seems normal, IMO.

i think this IR will become poison. 
%vecinit.i.i.i.i = insertelement <16 x i8> poison, i8 %conv.i.i.i, i64 0
am I right?

Reply:
If the BB is unreachable, it should be simplified by `SimplifyCFG`.

Reply:
> %vecinit.i.i.i.i = insertelement <16 x i8> poison, i8 %conv.i.i.i, i64 0
%vecinit15.i.i.i.i = shufflevector <16 x i8> %vecinit.i.i.i.i, <16 x i8> poison, <16 x i32> zeroinitializer

It is just a vector splat.
```
%vecinit.i.i.i.i = <16 x i8> <i8 %conv.i.i.i, i8 poison, i8 poison, ...>
%vecinit15.i.i.i.i = <16 x i8> <i8 %conv.i.i.i, i8 %conv.i.i.i, i8 %conv.i.i.i, ...>
```

Reply:
This does look suspicious. It's possible that the unreachability information was later lost (e.g. an assume false got dropped). Or it could be some bug in the reachability analysis...

Reply:
I'm on analyze it from now.

Reply:
@dtcxzyw can I know what optimize pass used? because when I use -passes=instcombine or -O1, handleUnreachableTerminator was not called. so, maybe I need some specific optimizing passes to trigger it .

Reply:
maybe I found why this mis-compile occurred.

Reply:
`store i32 1, ptr undef` is an indicator that the following code is unreachable, so this sounds correct to me. The question would be more why `store i32 1, ptr undef` later gets dropped -- it's supposed to be converted into `unreachable` by SimplifyCFG.

Reply:
Is this issue arising from something other than what I mentioned?

Reply:
I don't think I really understood what you meant...

Reply:
I wonder my assumption is correct I wrote at comment. `store i32 1, ptr undef` was not removed when I tested, beside it was crashed because `%ref` was removed as DCE in 2nd combining.

Reply:
Ah, when you say "crash", you mean the fixpoint verification failure? If so, yes, there is a problem in that we fail to report the reduced use-count when replacing the operand. Maybe it's best not to share a helper, so the code inside InstCombine can use replaceUse(), which does this correctly.

(Though I don't think this is really related to the miscompile -- if there even is a miscompile?)

Reply:
Emm, looks like it still exists. But I don't think we should block this patch :)